README.md
# js_paginate
Pagination extension provides the pagination for the custom extension records.


ChangeLog
2016-11-23  Jainish Senjaliya, <jainishsenjaliya@gmail.com>

	* Extension Manual Update
	* Modify composer file
	* Released version 3.0.3

2016-11-05  Jainish Senjaliya, <jainishsenjaliya@gmail.com>

	* Extension Manual Update
	* Released version 3.0.2

2016-09-29  Jainish Senjaliya, <jainishsenjaliya@gmail.com>

	* Added composer file
	* Released version 3.0.1

2016-09-14  Jainish Senjaliya, <jainishsenjaliya@gmail.com>

	* Added multilanguage files for br, cz, dk, es, fi, hr, hu, it, nl, no, ru, se and tr
	* Released version 3.0.0

2016-08-13  Jainish Senjaliya, <jainishsenjaliya@gmail.com>

	* Added language files for French
	* Update manual
	* Released version 2.0.7

2016-07-12  Jainish Senjaliya, <jainishsenjaliya@gmail.com>

	* Update manual
	* Released version 2.0.6

2016-07-09  Jainish Senjaliya, <jainishsenjaliya@gmail.com>

	* Add manual
	* Released version 2.0.5

2016-05-31  Jainish Senjaliya, <jainishsenjaliya@gmail.com>

	* Added language files for Dutch
	* Released version 2.0.4

2016-05-20  Jainish Senjaliya, <jainishsenjaliya@gmail.com>

	* Added dotted ( ... ) in pagination. if we reach more than maximum Visible Pages limit then dotted will be added after maximum Visible Pages
	* Released version 2.0.3

2016-05-06  Jainish Senjaliya, <jainishsenjaliya@gmail.com>

	* Added language files for German
	* Compatibility with TYPO3 8
	* Released version 2.0.2

2016-04-04  Jainish Senjaliya, <jainishsenjaliya@gmail.com>

	* Added fancy layout option and manage through typoscript
	* Released version 2.0.1

2016-03-27  Jainish Senjaliya, <jainishsenjaliya@gmail.com>

	* Compatibility with TYPO3 8.0
	* Released version 2.0.0

2016-03-25  Jainish Senjaliya, <jainishsenjaliya@gmail.com>

	* Added Basic layout. Manage through typoscript
	* Released version 1.0.4

2016-01-23  Jainish Senjaliya, <jainishsenjaliya@gmail.com>

	* Added First and last options in pagination
	* Released version 1.0.3

2015-12-23  Jainish Senjaliya, <jainishsenjaliya@gmail.com>

	* Bug Fixes for TYPO3 version 7
	* Released version 1.0.2

2015-12-19  Jainish Senjaliya, <jainishsenjaliya@gmail.com>

	* Compatibility with TYPO3 7.6
	* Released version 1.0.1

2015-11-19  Jainish Senjaliya, <jainishsenjaliya@gmail.com>

	* This is simple Extension for Pagination
	* We can add pagination before content and after content area
	* We can set number of items Per Page 
	* We can set maximum Visible Pages limit
	* Released version 1.0.0

Classes/Controller/PaginateController.php
<?php

namespace JS\JsPaginate\Controller;
 
/***************************************************************
 *  Copyright notice
 *
 *  (c) 2015-2016 Jainish Senjaliya <jainishsenjaliya@gmail.com>
 *
 *  All rights reserved
 *
 *  This script is part of the TYPO3 project. The TYPO3 project is
 *  free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  The GNU General Public License can be found at
 *  http://www.gnu.org/copyleft/gpl.html.
 *
 *  This script is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  This copyright notice MUST APPEAR in all copies of the script!
 ***************************************************************/

 
/**
 * This view helper uses the technology of paginate widget but works with arrays
 * and the assigned objects don't need the QueryResultInterface.
 *
 * @package Typo3
 *
 * @author   Jainish Senjaliya <jainishsenjaliya@gmail.com>
 * @license http://www.gnu.org/licenses/gpl.html GNU General Public License, version 3 or later
 */
class PaginateController extends \TYPO3\CMS\Fluid\Core\Widget\AbstractWidgetController {

	/**
	 * @var array
	 */
	protected $configuration = array('itemsPerPage' => 5, 'insertAbove' => FALSE, 'insertBelow' => TRUE, 'maximumVisiblePages' => 7, 'displayDots' => FALSE);

	/**
	 * @var array
	 */
	protected $objects;

	/**
	 * @var integer
	 */
	protected $currentPage = 1;

	/**
	 * @var integer
	 */
	protected $numberOfPages = 1;

	/**
	 * @var integer
	 */
	protected $itemsPerPage = 0;

	/**
	 * Initialize Action of the widget controller
	 *
	 * @return void
	 */
	public function initializeAction() {
		$this->objects = $this->widgetConfiguration['objects'];
		if (version_compare(TYPO3_branch, '6.2', '<')) {
			$this->configuration = \TYPO3\CMS\Core\Utility\GeneralUtility::array_merge_recursive_overrule($this->configuration, $this->widgetConfiguration['configuration'], TRUE);
		} else {
			if (version_compare(TYPO3_branch, '8.0', '<')) {
				$this->configuration = \TYPO3\CMS\Core\Utility\GeneralUtility::array_merge($this->configuration, $this->widgetConfiguration['configuration'], TRUE);
			}else{
				$this->configuration = array_merge($this->configuration, $this->widgetConfiguration['configuration']);
			}
		}
	}

	/**
	 * Returns the items per page
	 *
	 * @return integer the items per page
	 */
	public function getItemsPerPage() {
		return $this->itemsPerPage;
	}

	/**
	 * Sets the items per page
	 *
	 * @param integer $itemsPerPage the items per page
	 *
	 * @return void
	 */
	public function setItemsPerPage($itemsPerPage) {
		if (empty($itemsPerPage)) {
			$this->itemsPerPage = (integer)$this->configuration['itemsPerPage'];
		} else {
			$this->itemsPerPage = $itemsPerPage;
		}
	}

	/**
	 * Returns the number of pages
	 *
	 * @return integer
	 */
	public function getNumberOfPages() {
		return $this->numberOfPages;
	}

	/**
	 * Sets the number of pages
	 *
	 * @param integer $numberOfPages the number of pages
	 *
	 * @return void
	 */
	public function setNumberOfPages($numberOfPages) {
		$this->numberOfPages = $numberOfPages;
	}

	/**
	 * Returns the current page
	 *
	 * @return integer the current page
	 */
	public function getCurrentPage() {
		return $this->currentPage;
	}

	/**
	 * Sets the current page and limits it between 1 and $this->numberOfPages.
	 *
	 * @param integer $currentPage the current page
	 *
	 * @return void
	 */
	public function setCurrentPage($currentPage) {
		$this->currentPage = $currentPage;
		if ($this->currentPage < 1) {
			$this->currentPage = 1;
		} elseif ($this->currentPage > $this->numberOfPages) {
			$this->currentPage = $this->numberOfPages;
		}
	}

	/**
	 * Returns all visible objects within a range, depending on itemsPerPage and the currentPage.
	 *
	 * @param \TYPO3\CMS\ExtBase\Persistence\QueryResult|array $objects the list of objects
	 *
	 * @return array<mixed> the list of visible objects
	 */
	public function getVisibleObjects($objects) {
		$i = 0;
		$modifiedObjects = array();
		$indexMin = $this->itemsPerPage * ($this->currentPage - 1);
		$indexMax = $this->itemsPerPage * $this->currentPage - 1;

		foreach ($objects as $object) {
			if ($i >= $indexMin && $i <= $indexMax) {
				$modifiedObjects[] = $object;
			}
			$i++;
		}

		return $modifiedObjects;
	}

	/**
	 * Index action of the widget controller
	 *
	 * @param integer $currentPage
	 * @param integer $itemsPerPage
	 *
	 * @return void
	 */
	public function indexAction($currentPage = 1, $itemsPerPage = 0) {
		$this->setItemsPerPage($itemsPerPage);
		$this->setNumberOfPages(intval(ceil(count($this->objects) / (integer)$this->itemsPerPage)));
		$this->setCurrentPage((integer)$currentPage);

		$this->view->assign('contentArguments', array(
			$this->widgetConfiguration['as'] => $this->getVisibleObjects($this->objects)
		));
		$this->view->assign('configuration', $this->configuration);
		if($this->numberOfPages >= 2) {
			$this->view->assign('pagination', $this->buildPagination());		  
		}
		$this->view->assign('itemsPerPage', $this->itemsPerPage);
	}

	/**
	 * Returns an array with the keys "pages", "current", "numberOfPages", "nextPage" & "previousPage"
	 *
	 * @return array
	 */
	protected function buildPagination() {
		
		$sidePageCount = intval($this->configuration['maximumVisiblePages']) >> 1;

		$firstPage = max($this->currentPage - $sidePageCount, 1);
		$lastPage = min($firstPage + $sidePageCount * 2, $this->numberOfPages);
		$firstPage = max($lastPage - $sidePageCount * 2, 1);

		$pages = array();
		foreach (range($firstPage, $lastPage) as $index) {
			$pages[] = array(
				'number' => $index,
				'isCurrent' => ($index === $this->currentPage)
			);
		}

		$pagination = array(
			'pages' => $pages,
			'current' => $this->currentPage,
			'numberOfPages' => $this->numberOfPages,
			'itemsPerPage' => $this->itemsPerPage,
			'firstPage' => 1,
			'lastPage' => $this->numberOfPages,
			'isFirstPage' => ($this->currentPage == 1),
			'isLastPage' => ($this->currentPage == $this->numberOfPages),
			'hasLessPages' => $firstPage > 2,
			'hasMorePages' => $lastPage + 1 < $this->numberOfPages
		);

		return $this->addPreviousAndNextPage($pagination);
	}

	/**
	 * Adds the nextPage and the previousPage to the pagination array
	 *
	 * @param array $pagination the pagination array which get previous and
	 *	  next page
	 *
	 * @return array the pagination array which contains some meta data and
	 *	   another array which are the pages
	 */
	protected function addPreviousAndNextPage($pagination) {
		if ($this->currentPage < $this->numberOfPages) {
			$pagination['nextPage'] = $this->currentPage + 1;
		}
		if ($this->currentPage > 1) {
			$pagination['previousPage'] = $this->currentPage - 1;
		}

		return $pagination;
	}
}
?>

Classes/ViewHelpers/PaginateViewHelper.php
<?php
namespace JS\JsPaginate\ViewHelpers;
 
/***************************************************************
*  Copyright notice
*
*  (c) 2015-2016 Jainish Senjaliya <jainishsenjaliya@gmail.com>
*
*  All rights reserved
*
*  This script is part of the TYPO3 project. The TYPO3 project is
*  free software; you can redistribute it and/or modify
*  it under the terms of the GNU General Public License as published by
*  the Free Software Foundation; either version 2 of the License, or
*  (at your option) any later version.
*
*  The GNU General Public License can be found at
*  http://www.gnu.org/copyleft/gpl.html.
*
*  This script is distributed in the hope that it will be useful,
*  but WITHOUT ANY WARRANTY; without even the implied warranty of
*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
*  GNU General Public License for more details.
*
*  This copyright notice MUST APPEAR in all copies of the script!
***************************************************************/
 
/**
 * This widget is a copy of the fluid paginate widget. Now it's possible to
 * use arrays with paginate, not only query results.
 *
 * @author     Jainish Senjaliya <jainishsenjaliya@gmail.com>
 * @copyright  2015 Copyright belongs to the respective authors
 * @license    http://www.gnu.org/licenses/gpl.html GNU General Public License, version 3 or later
 */
 
class PaginateViewHelper extends \TYPO3\CMS\Fluid\Core\Widget\AbstractWidgetViewHelper {
 
    /**
     * @var \JS\JsPaginate\Controller\PaginateController
     */
    protected $controller;
 
    /**
     * Injection of widget controller
     *
     * @param \JS\JsPaginate\Controller\PaginateController $controller
     * @return void
     */
    public function injectController(\JS\JsPaginate\Controller\PaginateController $controller) {
        $this->controller = $controller;
    }
 
    /**
     * The render method of widget
     *
     * @param mixed $objects \TYPO3\CMS\ExtBase\Persistence\QueryResultInterface,
     *        \TYPO3\CMS\ExtBase\Persistence\ObjectStorage object or array
     * @param string $as
     * @param array $configuration
     * @return string
     */
    public function render($objects, $as, array $configuration = array('itemsPerPage' => 10, 'insertAbove' => FALSE, 'insertBelow' => TRUE)) {
        return $this->initiateSubRequest();
    }
}
 
?>

ext_emconf.php
<?php

/***************************************************************
 * Extension Manager/Repository config file for ext: "js_paginate"
 *
 * Auto generated by Extension Builder 2016-03-27
 *
 * Manual updates:
 * Only the data in the array - anything else is removed by next write.
 * "version" and "dependencies" must not be touched!
 ***************************************************************/

$EM_CONF[$_EXTKEY] = array(
	'title' => 'Paginate',
	'description' => 'Pagination extension provides the pagination for the custom extension records, News extension and other extensions',
	'category' => 'misc',
	'author' => 'Jainish Senjaliya',
	'author_email' => 'jainishsenjaliya@gmail.com',
	'state' => 'stable',
	'internal' => '',
	'uploadfolder' => '0',
	'createDirs' => '',
	'clearCacheOnLoad' => 0,
	'version' => '3.0.3',
	'constraints' => array(
		'depends' => array(
			'typo3' => '6.0.0 - 8.9.99',
		),
		'conflicts' => array(
		),
		'suggests' => array(
		),
	),
);